### ISSUE TITLE
Main NEWS file is not generated properly anymore

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6354]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)| @therve reported|
|-|-|
|Trac ID|trac#6354|
|Type|defect|
|Created|2013-03-06 09:52:09Z|
|Branch|https://github.com/twisted/twisted/tree/proper-news-delete-6354|

In [#4315](https://github.com/twisted/twisted/issues/4315), we introduced automatic removal of NEWS fragments. Unfortunately, that means that the subprojects NEWS files are built properly, but not the main one, as the fragments are deleted when it tries to build it. I'll attempt to fix it if it's easy, otherwise revert the commit if I can't.

<details><summary>Searchable metadata</summary>

```
trac-id__6354 6354
type__defect defect
reporter__therve therve
priority__high high
milestone__Twisted_13_0_0 Twisted-13.0.0
branch__branches_proper_news_delete_6354 branches/proper-news-delete-6354
branch_author__therve therve
status__closed closed
resolution__fixed fixed
component__release_management release management
keywords__ 
time__1362563529000000 1362563529000000
changetime__1362593069000000 1362593069000000
version__None None
owner__therve therve
cc__radix
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_2"></a>|@therve commented|
|-|-|

(In [37421]) Branching to 'proper-news-delete-6354'

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_3"></a>|@therve removed owner|
|-|-|

It's ready for review, fairly easy fix. I'm sad we didn't have the test before...

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_4"></a>|@therve commented|
|-|-|

Build results here: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/proper-news-delete-6354

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_5"></a>|@tomprince set owner to @therve|
|-|-|

It feels like too much logic is being put in `_iterprojects`. It seems like it might make more sense for it to just yield `topfiles, name, version` and then `buildAll` can build both the project and aggregate `NEWS` and then delete the files.

Otherwise, in isolation, the changes look good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @tomprince|
|-|-|

It sounds reasonable, changed in r37426.

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_7"></a>|@tomprince set owner to @therve|
|-|-|

  1. Add a note to _deleteFragments that it needs to be called from an svn repo.
  2. Calculate 
```python
"Twisted %s %s (%s)" % (name, version.base(), today)
```
     once.
Please merge after making the above changes.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve set status to `closed`|
|-|-|

(In [37437]) Merge proper-news-delete-6354

Author: therve
Reviewer: tom.prince
Fixes: [#6354](https://github.com/twisted/twisted/issues/6354)

Only delete NEWS fragment once all NEWS file are built, to prevent issues when
creating the main NEWS aggregating the others.
